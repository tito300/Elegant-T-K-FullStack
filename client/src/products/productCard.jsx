import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'
import propType from 'prop-types';
import Button from '../common/button';


class ProductCard extends Component {

    state = {
        success: null,
    }

    static propType = {
        item: propType.object.isRequired,
        i: propType.number.isRequired,
    }


    addItemToCart = (event) => {
        this.setState({ success: 'added' })
        setTimeout(() => {
            this.setState({ success: null });
        }, 2000);
    }

    handleProductClick = (e) => {
        this.props.history.push(`/product/${this.props.item.id}`)
    }

    render() {
        return (
            <div key={this.props.item.id}
                className={`products__card card-${this.props.i}`}>
                <div onClick={this.handleProductClick} className="products__card__img-cont">
                    <img src={this.props.item.photo} alt="" className="card-img"></img>
                </div>
                <div className="products__card__details">
                    <h3 onClick={this.handleProductClick} className="products__card__details__title">{this.props.item.title}</h3>
                    <p className="products__card__details__disc">{this.props.item.discreption}</p>
                </div>
                <div className="products__card__btn">
                    <p className="products__card__btn__price">${this.props.item.price}</p>
                    <Button
                        id={this.props.item.id}
                        text='View Item'
                        to={`/product/${this.props.item.id}`} />
                </div>
                <div className={`flash-success ${this.state.success}`}>
                    <p className="flash-s-msg">Item added to cart</p>
                </div>
            </div>
        )
    }
}


export default withRouter(ProductCard);