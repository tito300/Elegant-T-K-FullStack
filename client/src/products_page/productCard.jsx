import React, { Component } from 'react';
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux';
import AddToCartButton from '../common/addToCart-button'


class ProductCard extends Component {

    state = {
        success: null,
    }


    addItemToCart = (event) => {
        this.setState({ success: 'added' })
        setTimeout(() => {
            this.setState({ success: null });
        }, 2000);
    }

    handleProductClick = (e) => {
        this.props.history.push(`/product/${this.props.item.id}`)
    }

    render() {
        return (
            <div key={this.props.item.id}
                className={`products__card card-${this.props.i}`}>
                <div onClick={this.handleProductClick} className="products__card__img-cont">
                    <img src={this.props.item.photo} alt="" className="card-img"></img>
                </div>
                <div className="products__card__details">
                    <h3 onClick={this.handleProductClick} className="products__card__details__title">{this.props.item.title}</h3>
                    <p className="products__card__details__disc">{this.props.item.discreption}</p>
                </div>
                <div className="products__card__btn">
                    <p className="products__card__btn__price">${this.props.item.price}</p>
                    {/* this button below does not work currently because component was changed and now requires
                        options to be passed in. this component should be replaced with view Item button instead. */}
                    <AddToCartButton addItemToCartSuccess={this.addItemToCart} item={this.props.item} />
                </div>
                <div className={`flash-success ${this.state.success}`}>
                    <p className="flash-s-msg">Item added to cart</p>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        loggedin: state.user.active,
        cartItems: state.cart.cartItems,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addItemToCart: (id) => {
            /* 
             * this first dispatch returns a function instead of an object so that the thunk middleware
             * would recognize it and pass the get state object to it. this way reducers can share 
             * state when needed.
             * */
            dispatch((dispatch, getState) => {
                dispatch({ type: 'ADD_ITEM_TO_CART', id: id, products: getState().products.products });
            }
            );

        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(ProductCard));