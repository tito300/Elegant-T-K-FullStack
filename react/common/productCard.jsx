import React, { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';

class ProductCard extends Component {

    componentDidUpdate(prevProps, prevState) {
        /* used header to sync cart instead of this here - to be removed  */
        // if (this.props.i < 1) {
        //     if (this.props.loggedin) {
        //         axios(`/api/users/cart/updateCart`, {
        //             method: 'post',
        //             headers: { Authorization: `bearer ${localStorage.getItem('jwt')}` },
        //             data: {
        //                 items: this.props.cartItems,
        //             }
        //         })
        //             .then((res) => console.log('synched cart successfully'));
        //     }
        // }

    }

    addItemToCart = (event) => {
        event.preventDefault();
        const id = event.target.dataset.id;
        // this.props.addItemToCart(id)
        // should be implemented when backend is ready
        this.props.addItemToCart(id);
    }

    render() {
        return (
            <div key={this.props.item.id}
                className={`products__card card-${this.props.i}`}>
                <div className="products__card__img-cont">
                    <img src={this.props.item.photo} alt="" className="card-img"></img>
                </div>
                <div className="products__card__details">
                    <h3 className="products__card__details__title">{this.props.item.title}</h3>
                    <p className="products__card__details__disc">{this.props.item.discreption}</p>
                </div>
                <div className="products__card__btn">
                    <p className="products__card__btn__price">${this.props.item.price}</p>
                    <a href="#" onClick={this.addItemToCart} className="add-btn" data-id={this.props.item.id}>add to cart</a>
                </div>
                <div className="flash-success">
                    <p className="flash-s-msg">Item added to cart</p>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        loggedin: state.user.active,
        cartItems: state.cartItems,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addItemToCart: (id) => {
            dispatch({ type: 'ADD_ITEM_TO_CART', id: id })
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);