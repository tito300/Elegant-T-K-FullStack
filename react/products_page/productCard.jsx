import React, { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';

class ProductCard extends Component {
    // constructor(props) {
    //     super(props);
    //     this.flash = React.createRef();
    // }
    state = {
        success: null,
    }

    addItemToCart = (event) => {
        event.preventDefault();
        const id = event.target.dataset.id;
        this.props.addItemToCart(id);
        this.setState({ success: 'added' })
        setTimeout(() => {
            this.setState({ success: null });
        }, 2000);
    }

    render() {
        return (
            <div key={this.props.item.id}
                className={`products__card card-${this.props.i}`}>
                <div className="products__card__img-cont">
                    <img src={this.props.item.photo} alt="" className="card-img"></img>
                </div>
                <div className="products__card__details">
                    <h3 className="products__card__details__title">{this.props.item.title}</h3>
                    <p className="products__card__details__disc">{this.props.item.discreption}</p>
                </div>
                <div className="products__card__btn">
                    <p className="products__card__btn__price">${this.props.item.price}</p>
                    <a href="#" onClick={this.addItemToCart} className="add-btn" data-id={this.props.item.id}>add to cart</a>
                </div>
                <div className={`flash-success ${this.state.success}`}>
                    <p className="flash-s-msg">Item added to cart</p>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
    return {
        loggedin: state.user.active,
        cartItems: state.cart.cartItems,
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        addItemToCart: (id) => {
            /* 
             * this first dispatch returns a function instead of an object so that the thunk middleware
             * would recognize it and pass the get state object to it. this way reducers can share 
             * state when needed.
             * */
            dispatch((dispatch, getState) => {
                dispatch({ type: 'ADD_ITEM_TO_CART', id: id, products: getState().products.products });
            }
            );

        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductCard);