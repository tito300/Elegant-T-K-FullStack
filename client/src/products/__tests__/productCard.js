import { shallow } from 'enzyme';
import React from 'react';
import { ProductCard } from '../productCard';
import { createMemoryHistory } from 'history';

let wrapper;

describe('Products components', ()=>{
    beforeAll(()=>{
        let item = {
            title: "random",
            id: '23442',
            discreption: 'sdsdsdsd',
            photo: './img/ssdsdq.jpg'
        }
        let history = createMemoryHistory();

        // they generate on spot and cause problems with snapshot
        delete history.location.key;
        delete history.entries[0].key;

        wrapper = shallow(<ProductCard item={item} i='random' history={history} />);
    })

    it('should render productCard', ()=>{
    
        expect(wrapper).toMatchSnapshot();
    });

    it('should push history when clicked', ()=>{
        wrapper.find('.products__card__img-cont').simulate('click');
        
        expect(wrapper.instance().props.history.action).toBe('PUSH');
    })
})


