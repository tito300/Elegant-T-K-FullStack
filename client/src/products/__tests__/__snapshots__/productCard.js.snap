// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products components should render productCard 1`] = `
<div
  className="products__card card-random"
  key="23442"
>
  <div
    className="products__card__img-cont"
    onClick={[Function]}
  >
    <img
      alt=""
      className="card-img"
      src="./img/ssdsdq.jpg"
    />
  </div>
  <div
    className="products__card__details"
  >
    <h3
      className="products__card__details__title"
      onClick={[Function]}
    >
      random
    </h3>
    <p
      className="products__card__details__disc"
    >
      sdsdsdsd
    </p>
  </div>
  <div
    className="products__card__btn"
  >
    <p
      className="products__card__btn__price"
    >
      $
    </p>
    <_default
      id="23442"
      text="View Item"
      to="/product/23442"
    />
  </div>
</div>
`;
